sound/soc/sof/intel/hda-codec.c

```diff
@@ -37,16 +38,55 @@ static void hda_codec_load_module(struct hda_codec *codec)
 static void hda_codec_load_module(struct hda_codec *codec) {}
 #endif
 
+/* check jack status after resuming from suspend mode */
+void hda_codec_jack_check(struct snd_sof_dev *sdev, int status)
+{
+	struct sof_intel_hda_dev *hda = sdev->pdata->hw_pdata;
+	struct hda_bus *hbus = sof_to_hbus(sdev);
+	struct hdac_bus *bus = sof_to_bus(sdev);
+	struct hda_codec *codec;
+	int mask;
+
+	/*
+	 * there are two reasons for runtime resume
+	 * (1) waken up by interrupt triggered by WAKEEN feature
+	 * (2) waken up by pm get functions for some audio operations
+	 * For case (1), the bits in status mean which codec triggers
+	 * the interrupt and jacks will be checked on these codecs.
+	 * For case (2), we need to check all the non-hdmi codecs for some
+	 * cases like playback with HDMI or capture with DMIC. In these
+	 * cases, only controller is active and codecs are suspended, so
+	 * codecs can't send unsolicited event to controller. The jack polling
+	 * operation will activate codecs and unsolicited event can work
+	 * even codecs become suspended later.
+	 */
+	mask = status ? status : hda->hda_codec_mask;
+
+	list_for_each_codec(codec, hbus)
+		if (mask & BIT(codec->core.addr))
```

Comment:
@RanderWang @plbossart Sorry, I don't understand this. Now you check jack-status of every codec on every runtime-resume?.. This seems wrong to me. Jack status change should be detected *when* the change happens, not when the next rtpm-resume is happening. This really doesn't look right to me, please, explain.

Reply:
please check comments in this function